Refactor and implement views

DarkSide666 opened the pull request:
![screenshot](https://user-images.githubusercontent.com/1969119/51032299-2884ed80-15a8-11e9-9329-6d2c300d4be2.png)

Demo in /demos folder.

fix #23

romaninsh commented:
Love it!

romaninsh commented on demos/index.php:
```diff
@@ -0,0 +1,56 @@
+
+// create model for form
+$m2 = clone $m;
+if (isset($_GET['model_id'])) {
```
if($id = $app->stickyGet('model_id')) {

..

DarkSide666 replied:
addressed

romaninsh commented on src/Controller.php:
```diff
@@ -49,85 +87,118 @@ public function setUp(\atk4\data\Model $m)
 
+        //Imants: deprecated - use $m->audit_log_controller->audit_log_stack[0] instead
+        // or $m->audit_log_controller->custom_action and custom_fields properties in your beforeSave hook
+        //$m->audit_log = $a;
```
why not keep a reference to `audit_log` ?

what if there are multiple audit logs?

DarkSide666 replied:
I intentionally didn't want developer to mess directly with audit data model. He should use audit controller for that and audit controller should take care of audit data model.
I think that's better work-flow.

romaninsh commented on template/audit-lister.html:
```diff
@@ -0,0 +1,78 @@
+                {changes}
+                    {change}
+
+                    <!-- option 1 - show as segment with field name in header and labels inside -->
```
remove comments from {change}. Template is dumb and will add LOTS of those comments into output. 

Placing them between {changes} and {change} is best.

DarkSide666 replied:
fixed

romaninsh commented on template/audit-history.pug:
```diff
@@ -0,0 +1,2 @@
+| <{_element}div{/} id="{$_id}" class="{_ui}ui{/} {$class} {_class}{/}" style="{$style}" {$attributes}>{$Content}
```
why first char is pipe? See: https://github.com/atk4/ui/blob/develop/template/semantic-ui/element.pug

DarkSide666 replied:
fixed

romaninsh commented on template/audit-lister.pug:
```diff
@@ -0,0 +1,59 @@
+      .extra.text
+        | {changes}
+        | {change}
+        // option 1 - show as segment with field name in header and labels inside
```
remove comments from {change}. Template is dumb and will add LOTS of those comments into output.

Placing them between {changes} and {change} is best.

DarkSide666 replied:
fixed

romaninsh commented on tests/BasicTest.php:
```diff
@@ -1,15 +0,0 @@
-     */
-    public function testTesting()
-    {
-        $this->assertEquals('foo', 'foo');
```
ah, that was a good test 😆

DarkSide666 replied:
yeah - 100% coverage 👯‍♂️

romaninsh commented on tests/CustomTest.php:
```diff
@@ -19,14 +22,19 @@ function init()
-                $m->audit_log['action'] = 'genderbending';
+                //$m->audit_log['action'] = 'genderbending'; // deprecated usage
+                //$m->audit_log_controller->audit_log_stack[0]['action'] = 'genderbending';
+                $m->audit_log_controller->custom_action = 'genderbending';
```
can this be done through a method call? 

$m->addHistoryComment("genderbending");

DarkSide666 replied:
well - if that's in beforeSave hook, then it already have done audit->push() and set action='update'. so to change that at this point you have only one option - set `controller->custom_action` and/or `controller->custom_fields` properties which are respected in audit->pull().

I intentionally didn't want developer to mess directly with audit data model. He should use audit controller for that and audit controller should take care of audit data model.

DarkSide666 commented:
@romans you can review https://github.com/atk4/audit/pull/24 again
• addressed your comments
• refactored migrator and demo now use Admin layout
• still don't see why we would need $model->audit_log if we have $model->auditController (see comment)
• fixed diff value of log record when deleting record from model
• renamed audit_log_controller to auditController
• added more tests for auditing of nested actions - see MultiModelTests
• not fixed (can't be fixed) - audit still will create empty log record even if there was no changes in data model. See https://github.com/atk4/audit/issues/17 (edited) 

New look:
![2](https://user-images.githubusercontent.com/1969119/51047762-ed97af80-15d1-11e9-9cbe-38d9742761f6.png)
![1](https://user-images.githubusercontent.com/1969119/51047767-f25c6380-15d1-11e9-8042-d1acac0117ae.png)
